File: day6/NAre5/solution_part1.py
import sympy
import numpy as np
from pathlib import Path


file_name = 'input.txt'
file_path = Path(__file__).with_name(file_name)

internal_timers = np.fromfile(file_path, sep=',')

timers_0 = sympy.Matrix([np.count_nonzero(internal_timers == i)
                         for i in range(9)])

forawrd_timers = sympy.Matrix([[0, 1, 0, 0, 0, 0, 0, 0, 0],
                              [0, 0, 1, 0, 0, 0, 0, 0, 0],
                              [0, 0, 0, 1, 0, 0, 0, 0, 0],
                              [0, 0, 0, 0, 1, 0, 0, 0, 0],
                              [0, 0, 0, 0, 0, 1, 0, 0, 0],
                              [0, 0, 0, 0, 0, 0, 1, 0, 0],
                              [1, 0, 0, 0, 0, 0, 0, 1, 0],
                              [0, 0, 0, 0, 0, 0, 0, 0, 1],
                              [1, 0, 0, 0, 0, 0, 0, 0, 0]])

sum_fish = sympy.ones(1, timers_0.shape[0])


def fish_from_all_fish_after_phase(creation_phase):
    # timers_(x) = (forawrd_timers^x)*timers_0
    timers_x = (forawrd_timers**creation_phase)*timers_0

    num_of_fish = sum_fish*timers_x
    return list(num_of_fish)[0]


if __name__ == "__main__":
    print(fish_from_all_fish_after_phase(creation_phase=80))

File: day6/NAre5/solution_part1.js
const fs = require('fs')
const path = require('path')


file_name = 'input.txt'
file_path = path.join(__dirname, file_name)

const raw_data = fs.readFileSync(file_path, 'utf8')

const fish_from_fish_after_phase_cache = {}

const fish_initial_timers = raw_data.trim().split(',');

const creator_fish_default_timer = 7;
const created_fish_default_timer = 9;

const fish_from_fish_after_phase = (phase_len) => {
    if (phase_len <= 0)
        return 1

    if (fish_from_fish_after_phase_cache[phase_len])
        return fish_from_fish_after_phase_cache[phase_len];

    const num_of_fish =
        fish_from_fish_after_phase(phase_len - creator_fish_default_timer)
        + fish_from_fish_after_phase(phase_len - created_fish_default_timer);

    fish_from_fish_after_phase_cache[phase_len] = num_of_fish;

    return num_of_fish;
}

const fish_from_all_fish_after_phase = ({ creation_phase }) => {
    let fish_after_creations = 0;

    fish_initial_timers.forEach(fish_internal_timer => {
        fish_after_creations += fish_from_fish_after_phase(creation_phase - fish_internal_timer)
    });

    return fish_after_creations;
}

if (require.main === module) {
    console.log(fish_from_all_fish_after_phase({ creation_phase: 80 }));
}

module.exports = {
    fish_from_all_fish_after_phase
}
